### PR TITLE
feat: add quicktype tool

### PR DESCRIPTION
**Pull request by grant**
# Feat: Add quicktype tool

- Updates JSONSchema
  - `$id`: unique identifier for the schema
  - `title`: the human readable title
  - `name`: the name of this data type
- Moves registry tool in separate directory in `tools/`
- Inits TypeScript project, `tools/quicktype-wrapper/`
  - Uses [`google/gts`](https://github.com/google/gts) for basic automatic styling
  - Crawls a directory, and produces equivalent language files for every `schema.json` file.
  - Adds basic README for the tool.
- Tool can be run as a Node script or installed as a CLI.

## Example Output

```
***********************
** Quicktype Wrapper **
***********************
* Valid Languages (L): csharp,java,python,rust,crystal,ruby,golang,cplusplus,elm,swift,objective-c,typescript,javascript,kotlin,dart,pike,haskell
***********************
* Environment Variables:
- IN=/Users/timmerman/Documents/github/googleapis/google-cloudevents/proto/
- OUT=/Users/timmerman/Documents/out
- L=TYPESCRIPT
***********************

== START ==
- Finding all JSON Schemas (schema.json)...
Found 2 schema(s).
- LogEntryData                            : google/events/cloud/audit/v1/LogEntryData.ts
- MessagePublishedData                    : google/events/cloud/pubsub/v1/MessagePublishedData.ts
== END ==
```

### CODE REVIEW

**Comment by jskeet** [comment]
Will look at this on Tuesday (Monday is a public holiday in the UK).

**Comment by grant** [comment]
Requesting a review from Adam too.

File: tools/quicktype-wrapper/README.md [hunk 1]
```diff
@@ -2,16 +2,37 @@
...
+This tool can be installed as a global CLI or run as a Node script. Requires Node 8+.
+
+### CLI (Recommended)
+
+Install the CLI globally:
 
 ```sh
-npm run build
-npm start
+sudo npm run cli
```
Comment (grayside) [inline tools/quicktype-wrapper/README.md:14]: General practice is to never use sudo to install an npm packge. Why is this necessary?

File: tools/quicktype-wrapper/README.md [hunk 2]
```diff
@@ -2,16 +2,37 @@
...
+
+```sh
+npm run compile
+npm run start
+```
+
+## Run
+
+const IN = process.env.IN || `${HOMEDIR}/Documents/github/googleapis/google-cloudevents/proto/`;
+const OUT = process.env.OUT || `${HOMEDIR}/Documents/out`;
```
Comment (grayside) [inline tools/quicktype-wrapper/README.md:31]: Working in Documents seems odd? I assume github/googleapis/google-cloudevents/proto is coming from a separately git cloned repository?

File: tools/quicktype-wrapper/README.md (hunk 1 above)
Comment (grant) [reply tools/quicktype-wrapper/README.md:14]: This is necessary for some folks because you're installing a bin `qt` globally. Otherwise you'll see:

```
Error: EACCES: permission denied, access '/usr/local/lib/node_modules'
# or
Error: EACCES: permission denied, unlink '/usr/local/bin/qt'
```

File: tools/quicktype-wrapper/README.md (hunk 2 above)
Comment (grant) [reply tools/quicktype-wrapper/README.md:31]: These are some default values that if you don't specify one.
(I recommend we always specify one)

It's nice being able to just execute `qt`.

We could consider not having defaults or just writing an error, but working in `Documents` rather than cluttering ~ is better imo.

Comment (jskeet) [reply tools/quicktype-wrapper/README.md:31]: Is it normal in Node to use environment variables like this, rather than providing them as command-line arguments? In other environments I'd *definitely* be using command line arguments, and I wouldn't assume any defaults - it feels to me like the chances of the defaults being right for more than a single person are pretty small.

File: tools/quicktype-wrapper/README.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+To continually watch the script insead of building after every change,
+run this command in a separate terminal:
+
+```sh
+npm run watch
+```
+
+## Lint/Fix
+
+To fix style/lint mistakes, run:
```
Comment (jskeet) [inline tools/quicktype-wrapper/README.md:46]: What is this to fix mistakes in - the JSON schema files?

**Review by jskeet — COMMENTED** [review]
Just a couple of general comments. I'm not qualified to review the Typescript, but I'm happy enough with the general idea.

File: tools/quicktype-wrapper/README.md (hunk 3 above)
Comment (grant) [reply tools/quicktype-wrapper/README.md:46]: No, this README is in the `tools/quicktype-wrapper/` folder/project. The JSON schemas are not in this folder (in this repo, but really tangential).

Comment (jskeet) [reply tools/quicktype-wrapper/README.md:46]: So what *would* it fix? If this is just generating code, wouldn't any style/lint issues suggest a problem with the generator or the schema?

File: tools/quicktype-wrapper/README.md (hunk 2 above)
Comment (grant) [reply tools/quicktype-wrapper/README.md:31]: The goal of this tool is not to build a CLI to generate, it's to use the `npm run gen` script.

I think I can just remove these defaults.

File: tools/quicktype-wrapper/README.md (hunk 3 above)
Comment (grant) [reply tools/quicktype-wrapper/README.md:46]: Fixing the schema is out of scope of this PR/folder. Any tooling to do that would not live in the `tools/quicktype-wrapper` tool.

Why do you want this feature here?

Comment (jskeet) [reply tools/quicktype-wrapper/README.md:46]: I don't particularly - I'm trying to find out what this *does* fix mistakes in; that's what I've been asking.

Comment (grant) [reply tools/quicktype-wrapper/README.md]: Added a more detailed comment.

File: tools/quicktype-wrapper/README.md (hunk 2 above)
Comment (grant) [reply tools/quicktype-wrapper/README.md]: Removed defaults.

**Comment by grant** [comment]
Ready for another review. Hoping to merge soon.

File: tools/quicktype-wrapper/README.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+Here's an example:
+
+```sh
+IN=~/Documents/github/googleapis/google-cloudevents/proto
+OUT=~/Documents/out
+L=typescript
+```
+
+Then run `qt`, which will pick up these environment variables.
```
Comment (jskeet) [inline tools/quicktype-wrapper/README.md:40]: Again, I find it odd to use environment variables rather than command line arguments, but maybe that's just a Node convention.

**Review by jskeet — APPROVED** [review]
That's definitely clearer, thanks. I suspect I won't need to use this, at least for a while, so it makes sense to aim it at usability for you for now - and I'm definitely in favour of merging and iterating.

File: tools/quicktype-wrapper/README.md (hunk 1 above)
Comment (grant) [reply tools/quicktype-wrapper/README.md]: Removed it fyi.

File: tools/quicktype-wrapper/README.md (hunk 4 above)
Comment (grant) [reply tools/quicktype-wrapper/README.md:40]: Will create a GitHub issue.

**Comment by grant** [comment]
Filed https://github.com/googleapis/google-cloudevents/issues/61 per above question.
